core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -138,37 +138,141 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+    // 4. the last update time is no smaller than the old value.
+    // 4. the old partitions set is a subset of the new partitions set.
+    //
+    // plus, we should only try to update the metadata after the corresponding log entry has been successfully written and replicated (see TransactionStateManager#appendTransactionToLog)
+    //
+    // if valid, transition is done via overwriting the whole object to ensure synchronization
+
     val toState = pendingState.getOrElse(throw new IllegalStateException("Completing transaction state transition while it does not have a pending state"))
-    if (toState != newState) {
+
+    if (toState != newMetadata.state ||
+      producerId != newMetadata.producerId ||
+      txnLastUpdateTimestamp <= newMetadata.txnLastUpdateTimestamp) {
+
       false
     } else {
-      pendingState = None
-      state = toState
-      true
+      val updated = toState match {
+        case Empty => // from initPid
+          if (producerEpoch > newMetadata.producerEpoch ||
+            producerEpoch < newMetadata.producerEpoch - 1 ||
+            newMetadata.topicPartitions.nonEmpty ||
+            newMetadata.txnStartTimestamp != -1) {
+
+            false
+          } else {
+            txnTimeoutMs = newMetadata.txnTimeoutMs
+            producerEpoch = newMetadata.producerEpoch
+
+            true
+          }
+
+        case Ongoing => // from addPartitions
+          if (producerEpoch != newMetadata.producerEpoch ||
+            !topicPartitions.subsetOf(newMetadata.topicPartitions) ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            txnStartTimestamp < newMetadata.txnStartTimestamp) {
+
+            false
+          } else {
+            txnStartTimestamp = newMetadata.txnStartTimestamp
+            addPartitions(newMetadata.topicPartitions)
+
+            true
+          }
+
+        case PrepareAbort => // from endTxn
+          if (producerEpoch != newMetadata.producerEpoch ||
+            !topicPartitions.equals(newMetadata.topicPartitions) ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            txnStartTimestamp != newMetadata.txnStartTimestamp) {
+
+            false
+          } else {
+
+            true
+          }
+
+        case PrepareCommit => // from endTxn
+          if (producerEpoch != newMetadata.producerEpoch ||
+            !topicPartitions.equals(newMetadata.topicPartitions) ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            txnStartTimestamp != newMetadata.txnStartTimestamp) {
+
+            false
+          } else {
+
+            true
+          }
+
+        case CompleteAbort => // from write markers
+          if (producerEpoch != newMetadata.producerEpoch ||
+            newMetadata.topicPartitions.nonEmpty ||
+            txnTimeoutMs != newMetadata.txnTimeoutMs ||
+            newMetadata.txnStartTimestamp == -1) {
+
+            false
+          } else {
+            txnStartTimestamp = newMetadata.txnStartTimestamp
```

Comment:
I have not reviewed your expiration PR so my understanding might be inconsistent, but here is how I read the existing code: `txnStartTimestamp` is only set when there is an active txn ongoing, otherwise it should always be set to -1 (i.e. the above check on the new metadata). And expiration will simply skip the metadata when this field is -1.